src/core/model/table/dynamic_position_list_index.cpp

```diff
@@ -0,0 +1,176 @@
+                                                   size_t relation_size)
+    : clusters_(std::move(clusters)),
+      valid_records_number_(size),
+      next_record_id_(relation_size),
+      probing_table_cache_() {}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::CreateFor(
+        std::vector<ClusterValue>& records) {
+    ClusterCollection clusters;
+    for (size_t record_id = 0; record_id < records.size(); ++record_id) {
+        ClusterValue const& value = records[record_id];
+        clusters[value].push_back(record_id);
+    }
+
+    return std::make_unique<DynamicPositionListIndex>(std::move(clusters), records.size(),
+                                                      records.size());
+}
+
+void DynamicPositionListIndex::UpdateWith(
+        std::vector<std::pair<std::optional<size_t>, ClusterValue>>& inserted_records,
+        std::unordered_set<size_t> const& deleted_records_ids) {
+    if (inserted_records.empty() && deleted_records_ids.empty()) {
+        return;
+    }
+    ClusterCollection new_clusters;
+    for (auto const& [encoded_value_opt, cluster_value] : inserted_records) {
+        if (encoded_value_opt.has_value()) {
+            new_clusters[cluster_value].emplace_back(*encoded_value_opt);
+        } else {
+            new_clusters[cluster_value].emplace_back(next_record_id_++);
+            ++valid_records_number_;
+        }
+    }
+    valid_records_number_ -= deleted_records_ids.size();
+
+    for (auto it = clusters_.begin(); it != clusters_.end();) {
+        auto& [cluster_value, cluster] = *it;
+        ProcessDeleteToCluster(cluster, deleted_records_ids);
+        auto new_cluster_it = new_clusters.find(cluster_value);
+        if (new_cluster_it != new_clusters.end()) {
+            auto const& [new_cluster_value, new_cluster] = *new_cluster_it;
+            cluster.insert(cluster.end(), new_cluster.begin(), new_cluster.end());
+            new_clusters.erase(new_cluster_value);
+        }
+
+        if (cluster.empty()) {
+            it = clusters_.erase(it);
+        } else {
+            sort(cluster.begin(), cluster.end());
+            ++it;
+        }
+    }
+
+    for (auto& new_cluster : new_clusters) {
+        clusters_.emplace(std::move(new_cluster));
+    }
+}
+
+std::unordered_map<int, unsigned> DynamicPositionListIndex::CreateFrequencies(
+        Cluster const& cluster, std::vector<int> const& probing_table) {
+    std::unordered_map<int, unsigned> frequencies;
+
+    for (int index : cluster) {
+        frequencies[probing_table[index]]++;
+    }
+
+    return frequencies;
+}
+
+std::shared_ptr<std::vector<int> const> DynamicPositionListIndex::CalculateAndGetProbingTable()
+        const {
+    if (probing_table_cache_ != nullptr) return probing_table_cache_;
+    cluster_values_cache_.clear();
+    std::vector<int> probing_table = std::vector<int>(next_record_id_);
+    int next_cluster_id = 1;
+    cluster_values_cache_.push_back({});
```

Comment:
Explain what this does in a comment, or make it apparent from the nearby code.

Reply:
By "this" I mean the exact line this comment references. Why are we adding an empty value here?